### ISSUE TITLE
FEATURE: Bulkhead isolation policy

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#138]
# Proposal: Throttle policy
## Purpose
- To limit the number of concurrent parallel invocations through the policy.  
- To isolate groups of actions (bulkhead isolation) such that the impact of faulting calls is isolated to the resource-pool to which it is limited, and other threads/pools/capacity remain to continue serving other calls.
## Scope
- A **non-exception handling** **resilience** strategy.
- Support all sync and async variants
- Support non-result-returning invocations (`Action`) and result-returning (`Func<TResult>`)
- Support `Execute(...)` and `ExecuteAndCapture(...)` variants.  
#### Configuration syntax

``` cssharp
ThrottlePolicy throttlePolicy = Policy
  .ThrottleAsync(int maxParallelization[, int maxQueuedActions]);

ThrottlePolicy<TResult> throttlePolicy = Policy<TResult>
  .ThrottleAsync(int maxParallelization[, int maxQueuedActions]);
```
#### Execution syntax

```
// TResult async form 
TResult result = throttlePolicy
  .ExecuteAsync(Func<Task<CancellationToken, TResult>> executionFunc, CancellationToken token);

// void async form
throttlePolicy
  .ExecuteAsync(Func<CancellationToken, Task> executionFunc, CancellationToken token);
```
#### Implementation
-  a pair of `SemaphoreSlim`s
#### Operation
- Attempt to acquire a semaphore limited to maxParallelization + maxQueuedActions with no delay.
  - Throw `ThrottleQueueFullException : ExecutionRejectedException` if cannot acquire semaphore  immediately.
- Attempt to acquire a semaphore limited to maxParallelization (**UPDATE**: timeout from any wrapping timeout policy can be automatically handled by the `CancellationToken` passed in)
  - **EDIT** wait until can acquire semaphore (unless timed out)
- (once semaphore acquired) Execute `executionFunc`
- If `executionFunc` raises any kind of fault: rethrow / pass fault result to caller.  
#### Implementation and Operation - Semaphore form
- As for async.  
- **EDIT**: Edited to remove `TaskScheduler`/`ThreadPool` proposal, which will not be taken forward for v5.0.
## Naming

Possible names could include `Throttle`, `ParallelThrottle`, `MaxConcurrency`, `Isolate`?  Typical (_but not the only_) usage is bulkhead isolation; the effective operation (the undisputable thing) is to limit parallelization to a maximum concurrency.
# Comments?

Comments?  Alternative suggestions?  Extra considerations to bear in mind?  

Scenarios where you'd use this? (please share now - could inform development!)

**Comment by brunolauze** [comment]
My readings about Semaphore and TaskScheduler brings me to this exact conclusion : http://stackoverflow.com/questions/23550470/limit-parallelism-of-an-async-method-and-not-block-a-thread-pool-thread

While I did try with a QueuedTaskScheduler approch, I feel this is not the Api to accomplish proper throttling. But that's only my understanding. Please comments your thoughts on why TaskScheduler would be better.

I have push an idea on how the ThrottlePolicy could be shaped at [throttle-policy](https://github.com/brunolauze/Polly/commits/throttle-policy)

Review is appreciated. Thanks.

**Comment by reisenberger** [comment]
Hi @brunolauze. Many thanks for all your contributions and the article!

I think we are thinking along the same lines here: the StackOverflow article is very much in line with my original proposal.  I don't think a `TaskScheduler` or `QueuedTaskScheduler` is better: `Semaphore` and `QueuedTaskScheduler` rather have different (possible) roles to play in different scenarios.  

**Throttling async code:** `SemaphoreSlim` is definitely the way to go (your code, the Stack Overflow article, my original proposal all agree)

**Throttling sync code:** It depends what the purpose is, and if threads are being managed upstream.  

+ `SemaphoreSlim` a way to go (great lightweight way to go!) if threads are being managed upstream (indeed it _requires_ that something upstream from the `ThrottlePolicy` is placing the calls onto separate threads).  Also nice that implementation parallels async implementation.

+ A task scheduler (for example `QueuedTaskScheduler` in its thread-pool flavour) is one possible way to manage and schedule synchronous work on to a defined-size thread pool.  It's probably (per comments in the StackOverflow article) more suited to CPU-bound work.  Possible scenarios might be:
  - a 'listener' thread is receiving tasks, and wants to hand them off onto a parallel-limited pool of 'worker' threads to execute, with some of Polly's fault-handling built in.
  - CPU-bound work like processing audio / video / images uploaded by users.

The idea to offer both an `IsolationMode.ThreadPool` and `IsolationMode.Semaphore` for sync code was to give both options (do you want Polly to manage threads for you or not?), and based on [Hystrix offering both](https://github.com/Netflix/Hystrix/wiki/Configuration#executionisolationstrategy).  

That said, `IsolationMode.ThreadPool` feels the harder case / has some unresolved questions.  `QueuedTaskScheduler` is one way of providing exactly schedule-tasks-onto-a-fixed-threadpool (I have used in the past), but I haven't had time yet to evaluate the underlying implementation against the range of .NET standard platforms we want to support.  (Was in the proposal as a ? question-mark as it probably merited further thought before implementation!)

**Comment by reisenberger** [comment]
As well as the [first Hystrix link on ThreadPool versus Semaphore](https://github.com/Netflix/Hystrix/wiki/Configuration#executionisolationstrategy), Hystrix have an [extensive discussion here](https://github.com/Netflix/Hystrix/wiki/How-it-Works#isolation).  

(Some of the considerations can probably be seen differently in the 2016 .NET context, and recent Hystrix discussion suggests views evolving beyond what stated in those articles, but there's lots of food for thought in there!)

**Comment by reisenberger** [comment]
Another option would be to omit the `IsolationMode.ThreadPool` variant entirely - or initially. 

_Very interested in community feedback_: Do you see Use Cases for `IsolationMode.ThreadPool` throttling with Polly?

**Comment by reisenberger** [comment]
Many thanks @brunolauze again for your code both here and on Timeout.  I expect to be able to review in detail Sunday thru Wednesday.  

(I think we can take the `SemaphoreSlim` versions of Throttle forward independent of the `ThreadPool` version, which I/we all can think about in parallel.)

**Comment by reisenberger** [comment]
@BrunoLauze  Many thanks for getting stuck in to the `Timeout` and `Throttle` Policies!  

First feedback here to help us proceed `Throttle` (without any `ThreadPool` variant; will come back to that later).

Your `SemaphoreSlim` approach for the throttle is spot on.  Your code follows the retry pattern perfectly! But we can simplify some of that away, some aspects needed by retry we don’t need for Throttle:

+ The throttle policy is non-exception handling (won’t change its operation depending on specific exceptions: we’ll just let all exceptions propagate back for the caller (some other policy / the user) to deal).  Therefore:
  -  The Policy and Engine don’t need to take any exception predicates or result predicates.  
  - The syntax can all be static methods on `Policy`. ie ‘ThrottleSyntax/Async’ can contain `public partial class Policy` offering the configuration syntax you already have, as static methods on `Policy` (rather than extensions on `this PolicyBuilder policyBuilder`).  
  - The whole `catch` block is probably redundant (tho `finally` block looks spot on!)

+ The `policyStateFactory` is not relevant and can be removed.  Retry was a case where the scope of the state needed to be per execution (if configured for 3 retries, _each_ call to execute needs a fresh chance at 3 retries, so needs a fresh state instance to track this).  `Throttle` is more like `CircuitBreaker`.  The state (the number of calls already occupying the semaphore) wants to be _shared_ across calls through the policy, so the scope of the Semaphore is per-policy, not per-execution.  I would remove all the `ThrottleState` classes and move everything up to `ThrottlePolicy/Async` and `ThrottleEngine`.



+ _New info:_ In the async implementation, we must always honour continueOnCapturedContext, ie `await whatever.ConfigureAwait(continueOnCapturedcontext);` .  The default for libraries is `.ConfigureAwait(false)`, but in the context of needing to stay on a custom `TaskScheduler` as [for Orleans users](https://github.com/App-vNext/Polly/issues/49), they want `.ConfigureAwait(true)`.  So we decorate every `await` with `.ConfigureAwait(continueOnCapturedcontext);` . To avoid missing any, you can use [this great tool](https://github.com/aelij/ConfigureAwaitChecker) as a [plugin for ReSharper](https://resharper-plugins.jetbrains.com/packages/ConfigureAwaitChecker.v9), or apparently as a Roslyn code analyzer too (CA2007?), or just :eyeglasses: 

+ I possibly hadn’t explained the intended interaction of `maxParallelization` and `maxQueuedActions` clearly in the original proposal.  The intention is:

`maxParallelization` : the maximum number of actions that may be executing simultaneously: governed by a `SemaphoreSlim` that callers block on until they can acquire it or are cancelled (ie exactly as you have it @brunolauze)

`maxQueuedActions` : the maximum number of executions that may be queueing waiting for the semaphore.  So a user might configure `maxParallelization = 8` actions are allowed to be running in parallel, but `maxQueuedActions` = another 20 actions are allowed to be queueing, waiting to acquire the main semaphore, if it is full.  The default for `maxQueuedActions` if not supplied by policy config overload should be 0.  

The intended implementation was another SemaphoreSlim of capacity  `maxParallelization` + `maxQueuedActions`.  If that cannot be acquired immediately, the `SemaphoreQueueFullException` (custom exception we define) would be thrown.  One can try to acquire a semaphore immediately without waiting, using `.Wait(TimeSpan.Zero, CancellationToken)` ([reference](http://stackoverflow.com/questions/33139724/how-to-check-whether-a-semaphoreslim-can-be-entered); [reference](http://blog.stephencleary.com/2013/04/cancellationtoken-behavior-with-asyncex.html)).

@brunolauze Are you happy to proceed like this?  Is this helpful? You are doing a great job of getting stuck into the codebase!

**Comment by reisenberger** [comment]
@brunolauze I reflected further on the `TaskScheduler` approach and agree it should not form part of  `Throttle` now.

My inspiration was that we use `QueuedTaskScheduler` very successfully in house to handle incoming messages in a microservice environment.  The 'listener' is on a dedicated single thread, offloading messages to `Task`s for processing in parallel.  `QueuedTaskScheduler`'s parallelization-limiting features, queueing features (tasks can be queued but not executed until capacity is available), coupled with its prioritisation features, make a great combination here - we can balance message priorities such that messages involved in supporting immediate user interactions can be prioritised, for example, and not swamped by batch tasks or lower-priority operations.  We can use parallelization optimisations to create back pressure ... I was keen to offer some of this value in Polly.

The issue for Polly is that this 'offloading' (from a single thread onto multiple background threads/`Task`s) isn't an exact fit with either the existing `Execute()` or `ExecuteAsync()` syntax.  `Execute()` clearly is sync and expects to stay on the same thread.  The `ExecuteAsync()` syntax (with its emphasis on `continueOnCapturedContext`) is geared to the `async`/`await` use of `Task` rather than the TPL (Task Parallel Library) use.  

`TaskScheduler`s _can_ bind nicely with the throttling concept, then, per above example, but if we are to introduce `TaskScheduler`s to Polly, we might want to consider a new, differentiating syntax, perhaps something like `ScheduleAsync(...)`.  And further thought / debate would be needed on value versus complexity.

If other users are interested in getting `TaskScheduler`s into Polly, please comment here or (when this is closed) start a new issue!

**Comment by reisenberger** [comment]
**PROPOSAL**: Change the name of the `ThrottlePolicy`, for release, to `IsolationPolicy`.  Configuration syntax becomes: 

```
IsolationPolicy isolationPolicy = Policy
  .IsolateAsync(int maxParallelization[, int maxQueuedActions]);
```

**Why?**: 

+ Reactive Extensions for .NET already has a `.Throttle` [operator](https://msdn.microsoft.com/en-us/library/hh229400(v=vs.103).aspx).  If in future we build [Polly.Rx](https://github.com/App-vNext/Polly/issues/134), we would have a name collision on Throttle.  _Note_: Rx's Throttle does something quite different from Polly's: Rx's throttle limits results after CPU has already been spent generating them; Polly's throttle-isolation limits executions _before_ CPU/other resources are spent on them. 

+ While the _implementation_ is a parallelism throttle, the _resilience_ term for this approach is bulkhead isolation.  

Views?

**Comment by mfjerome** [comment]
I agree with the explanation to rename the policy. But just thinking out loud here, `Isolation` is quite a large term isn't it? What about simply Bulkhead?

**Comment by brunolauze** [comment]
In fact, 
Policy.Isolate() is in direct clash with 
var policy = Policy.CircuitBreaker()
policy.Isolate()
therefore I bring a lot more confusion even if we work around the naming clash.

Other suggestions might be better :
.Bulkhead()
.LimitConcurrency()
Let's have a look at this branch with the name changed:

https://github.com/brunolauze/Polly/tree/v5.0-alpha-isolate
What do you think?
Thanks.

Date: Thu, 15 Sep 2016 12:20:41 -0700

**Comment by brunolauze** [comment]
In fact,
Policy.Isolate() is in direct clash with
var policy = Policy.CircuitBreaker()
policy.Isolate()
therefore I bring a lot more confusion even if we work around the naming clash.

Other suggestions might be better :
.Bulkhead()
.LimitConcurrency()
Let's have a look at this branch with the name changed:

https://github.com/brunolauze/Polly/tree/v5.0-alpha-isolate
What do you think?

**Comment by reisenberger** [comment]
@brunolauze Good catch! (that we have another meaning for `Isolate` elsewhere).  @brunolauze @mfjerome I'll build on your suggestion of `Bulkhead` for now.

(Time for reflection still before release - _further community feedback welcome!_ - but `.Bulkhead()` seems leader of the pack for now ... :tiger:  )

**Comment by brunolauze** [comment]
@reisenberger @mfjerome  I think it's a good candidate because it follows directly <PatternName>Policy as RetryPolicy and CircuitBreakerPolicy

**Comment by reisenberger** [comment]
So: Bulkhead isolation policy has been delivered into the v5.0-alpha branch, ready for a forthcoming v5.0-alpha nuget release

**Comment by reisenberger** [comment]
_Commentary on .NET4.0 async support for Bulkhead Policy:_ `SemaphoreSlim` does not offer the `.WaitAsync(...)` async variant of `.Wait()`.  Therefore, the .NET4.0 implementation (_only_) of bulkhead is currently _blocking_, for `BulkheadPolicyAsync`.

This is in principle easily solvable with (internal `#if`/`#endif#` forking and) use of eg [Stephen Toub's async semaphore](https://blogs.msdn.microsoft.com/pfxteam/2012/02/12/building-async-coordination-primitives-part-5-asyncsemaphore/) or [Stephen Cleary's async semaphore](https://github.com/StephenCleary/AsyncEx/wiki/AsyncSemaphore).  At this point this work has not been undertaken, in case there are also .NET4.0 compatibility challenges in other v5.0 policies. A wider review of  .NET4.0 compatibility will be undertaken once all v5.0 policies are in place.
